### DISCUSSION TITLE
Setting up M2M relation on item creation fails

### DISCUSSION THREAD

**Discussion by kadlinobit** [directus/directus#10909] [Archive]
Hi, 
I think this might be a bug, but I'm not sure yet. I have tables `dj` and `genre` connected via conjuction table `dj_genre`.

![image](https://user-images.githubusercontent.com/46809506/148556227-e432989c-6602-43fc-bac1-b690ff59489e.png)

- The alias M2M field on `dj` table is called `genres`. 
- I have 2 records in the `genre` table with IDs `1` and `2`. 

Now I am creating a `dj` record with following payload:
```json
{
    "name": "Test DJ",
    "genres": [1,2]
}
```
Creation fails with Forbidden error:
```json
{
    "errors": [
        {
            "message": "You don't have permission to access this.",
            "extensions": {
                "code": "FORBIDDEN"
            }
        }
    ]
}
```
When I omit `genres` from the payload, creation of `dj` record is successful.

There is the same situation when trying to update existing `dj` record:

```json
{
    "name": "Updated dj name",
    "genres": [1]
}
```

But when I put it like this, it works:

```json
{
    "name": "Updated dj name",
    "genres": [{
        "genre_id": 1,
        "dj_id": "1617149b-6f9e-44a2-b494-df6799a78c09"
    }]
}
```

So IMHO the problem is that simple syntax `"genres" : [1]` is not expecting an existing `genre` record ID, but rather existing `dj_genre` (conjuction table) record ID. And because there is no `dj_genre` record with ID = 1, the request fails. 
Is that a desired behavior? Because it makes logical sense, but it's actually a non-sense from the user / developer point of view. No one is interested in the conjunction table record ID. 

Furthermore, it's impossible to create `dj_genre` connections on `dj` record insertion, because we don't have a `dj` ID yet.

**Comment by kadlinobit** [comment]
OK I think I got it. When creating a new `dj` record, I can set up the relationship to `genre` like this:
```json
{
    "name": "Newly created DJ Name",
    "genres": [{"genre_id": 1}, {"genre_id": 2}]
}
```
Which means that I don't have to provide `dj_id` value for the junction table - it is filled automatically. Good to know! The API docs on managing relations should also provide some examples for working with M2M. I know it's just a special case of double-sided one-to-many (O2M => junction <= M2O), but I was intuitively expecting to be able to refer directly to the relation table and not to the junction one. 
https://docs.directus.io/reference/introduction/#relational-data

**Reply by kadlinobit** [reply]
Still I'd like to know the best way to handle for example updates of `dj` record without having to do some difference calculations on selected `genres` in frontend. 
When I send an update of a `dj` record, that already has genres 1 and 2 linked, with following payload:
```json
{
    "genres": [{"genre_id": 1}, {"genre_id": 2}]
}
```
Directus will create 2 new records in the junction table - and for the already existing junction records (linking `dj` with the same genres), it will remove the `dj_id` - keeping these "orphaned" records in the database.
![image](https://user-images.githubusercontent.com/46809506/148656179-59cb1431-7509-47ea-a46b-8e6bcd99aa3c.png)

**Reply by rijkvanzanten** [reply]
@kadlinobit There's a setting in the relations for the field that allow you to control whether the item should be orphaned or deleted:

<img width="398" alt="CleanShot 2022-01-11 at 17 31 36@2x" src="https://user-images.githubusercontent.com/9141017/149031920-8f5d72de-2982-4cd0-8f76-9a000d3fba57.png">

Other than that, you can include the primary key of the junction table to have the API update it instead of create a new one, eg:

```json
{
    "genres": [{"id": 10, "genre_id": 1}, {"id": 11, "genre_id": 2}]
}
```

**Reply by kadlinobit** [reply]
@rijkvanzanten Wow I missed that setting, thanks a ton! I'm really impressed, everytime I hit some problem, it turns out that Directus has it covered :)

**Reply by benhaynes** [reply]
17 years of product development will do that! 😂
